From 060e7c615b5730d3ca309735700489863e0a4ddd Mon Sep 17 00:00:00 2001
From: sajcho <saux.aarch64@gmail.com>
Date: Fri, 9 Dec 2022 20:40:25 +0100
Subject: [PATCH] resolvconf.in: fix - we have sv installed in the sbin
 directory

---
 resolvconf.in | 12 ++++++------
 1 file changed, 6 insertions(+), 6 deletions(-)

diff --git a/resolvconf.in b/resolvconf.in
index a88fce3..45817c6 100644
--- a/resolvconf.in
+++ b/resolvconf.in
@@ -347,12 +347,12 @@ detect_init()
 			then
 				/usr/sbin/service \$1 restart
 			fi"
-	elif [ -x /bin/sv ]; then
-		RESTARTCMD='/bin/sv status $1 >/dev/null 2>&1 &&
-			    /bin/sv try-restart $1'
-	elif [ -x /usr/bin/sv ]; then
-		RESTARTCMD='/usr/bin/sv status $1 >/dev/null 2>&1 &&
-			    /usr/bin/sv try-restart $1'
+	elif [ -x /sbin/sv ]; then
+		RESTARTCMD='/sbin/sv status $1 >/dev/null 2>&1 &&
+			    /sbin/sv try-restart $1'
+	elif [ -x /usr/sbin/sv ]; then
+		RESTARTCMD='/usr/sbin/sv status $1 >/dev/null 2>&1 &&
+			    /usr/sbin/sv try-restart $1'
 	elif [ -e /etc/arch-release ] && [ -d /etc/rc.d ]; then
 		RCDIR=/etc/rc.d
 		RESTARTCMD='
-- 
2.38.1

