From 2dad0072ec664c68d9344a8f9ef034c3d1e08c73 Mon Sep 17 00:00:00 2001
From: sajcho <saux.aarch64@gmail.com>
Date: Wed, 15 Feb 2023 23:12:32 +0100
Subject: [PATCH] Makefile.in: modified by saux

---
 Makefile.in | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/Makefile.in b/Makefile.in
index bce64e7..405e9f1 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -491,6 +491,10 @@ dist-xz: distdir
 	tardir=$(distdir) && $(am__tar) | XZ_OPT=$${XZ_OPT--e} xz -c >$(distdir).tar.xz
 	$(am__post_remove_distdir)
 
+dist-zstd: distdir
+	tardir=$(distdir) && $(am__tar) | zstd -c  >$(distdir).tar.zst
+	$(am__post_remove_distdir)
+
 dist-tarZ: distdir
 	tardir=$(distdir) && $(am__tar) | compress -c >$(distdir).tar.Z
 	$(am__post_remove_distdir)
@@ -521,6 +525,8 @@ distcheck: dist
 	  lzip -dc $(distdir).tar.lz | $(am__untar) ;;\
 	*.tar.xz*) \
 	  xz -dc $(distdir).tar.xz | $(am__untar) ;;\
+	*.tar.zst*) \
+	  zstd -dc $(distdir).tar.zst | $(am__untar) ;;\
 	*.tar.Z*) \
 	  uncompress -c $(distdir).tar.Z | $(am__untar) ;;\
 	*.shar.gz*) \
@@ -704,7 +710,7 @@ uninstall-am:
 .PHONY: $(RECURSIVE_CLEAN_TARGETS) $(RECURSIVE_TARGETS) CTAGS GTAGS \
 	all all-am am--refresh check check-am clean clean-cscope \
 	clean-generic cscope cscopelist cscopelist-recursive ctags \
-	ctags-recursive dist dist-all dist-bzip2 dist-gzip dist-lzip \
+	ctags-recursive dist dist-all dist-bzip2 dist-gzip dist-lzip dist-zstd \
 	dist-shar dist-tarZ dist-xz dist-zip distcheck distclean \
 	distclean-generic distclean-hdr distclean-tags distcleancheck \
 	distdir distuninstallcheck dvi dvi-am html html-am info \
-- 
2.39.1

From a413284e9090f38150842de4fbd84fed34b3e59c Mon Sep 17 00:00:00 2001
From: sajcho <saux.aarch64@gmail.com>
Date: Fri, 17 Feb 2023 10:55:24 +0100
Subject: [PATCH] installtransaction.cpp: define supported architecture

---
 src/installtransaction.cpp | 17 ++++++++++++++++-
 1 file changed, 16 insertions(+), 1 deletion(-)

diff --git a/src/installtransaction.cpp b/src/installtransaction.cpp
index 59834f1..41f2a51 100644
--- a/src/installtransaction.cpp
+++ b/src/installtransaction.cpp
@@ -34,6 +34,21 @@ using namespace std;
 #include "lockfile.h"
 #endif
 
+#ifdef linux
+#ifdef __x86_64__
+#define PKG_EXT ".x86_64.pkg.tar."
+#endif
+#ifdef __aarch64__
+#define PKG_EXT ".aarch64.pkg.tar."
+#endif
+#ifdef __riscv
+#define PKG_EXT ".riscv64.pkg.tar."
+#endif
+#else
+#error "unsupported architecture"
+#endif
+
+
 using namespace StringHelper;
 
 
@@ -371,7 +386,7 @@ InstallTransaction::installPackage( const Package* package,
             args +=
                 package->name()    + "#" +
                 package->version() + "-" +
-                package->release() + ".pkg.tar." + getPkgmkCompressionMode();
+                package->release() + PKG_EXT + getPkgmkCompressionMode();
 
 
             // - inform the user about what's happening
-- 
2.39.1

