From 3d70356c09d0612ccdf88ae1e01126039e25e76f Mon Sep 17 00:00:00 2001
From: sajcho <saux.aarch64@gmail.com>
Date: Fri, 8 Sep 2023 10:35:36 +0200
Subject: [PATCH] dist.rs: change dir libexec to lib

---
 src/bootstrap/dist.rs | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/bootstrap/dist.rs b/src/bootstrap/dist.rs
index 32da4ac29a4..78a5faeae04 100644
--- a/src/bootstrap/dist.rs
+++ b/src/bootstrap/dist.rs
@@ -423,7 +423,7 @@ fn prepare_image(builder: &Builder<'_>, compiler: Compiler, image: &Path) {
                 },
                 builder.kind,
             ) {
-                builder.install(&ra_proc_macro_srv, &image.join("libexec"), 0o755);
+                builder.install(&ra_proc_macro_srv, &image.join("lib"), 0o755);
             }
 
             let libdir_relative = builder.libdir_relative(compiler);
-- 
2.42.0

